src/modules/flow/form/form.c

```diff
@@ -1345,5 +1363,572 @@ integer_enabled_set(struct sol_flow_node *node,
     return 0;
 }
 
+struct integer_custom_data {
+    struct integer_data base;
+    struct sol_timeout *timer;
+    char *chars;
+    size_t cursor_row, cursor_col, value_prefix_len;
+    int blink_time;
+    uint8_t n_digits;
+    bool blink_on : 1;
+    bool state_changed : 1;
+    bool cursor_initialized : 1;
+};
+
+static int
+integer_custom_format_do(struct sol_flow_node *node)
+{
+    struct integer_custom_data *mdata = sol_flow_node_get_private_data(node);
+    int r, buf_size = mdata->base.text_grid.capacity;
+    bool no_more_space = false;
+    size_t row = 0, col = 0, tmp;
+    char *it_value = NULL;
+    char format[PATH_MAX];
+
+    if (!mdata->state_changed) {
+        char *value;
+        size_t pos;
+
+        if (mdata->cursor_col > mdata->base.columns - 1)
+            goto send;
+
+        pos = coords_to_pos
+                (mdata->base.columns, mdata->cursor_row, mdata->cursor_col);
+
+        value = (char *)sol_buffer_at(&mdata->base.text_grid, pos);
+
+        if (mdata->blink_on) {
+            mdata->blink_on = false;
+            *value = SPC;
+        } else {
+            mdata->blink_on = true;
+            pos -= coords_to_pos(mdata->base.columns, mdata->cursor_row, 0);
+            pos -= mdata->value_prefix_len;
+            *value = mdata->chars[pos];
+        }
+
+        goto send;
+    }
+
+    r = format_title(&mdata->base.text_grid, buf_size, mdata->base.rows,
+        mdata->base.columns, &row, &col, mdata->base.format, mdata->base.title,
+        mdata->base.title_tag, mdata->base.value_tag, &no_more_space);
+    SOL_INT_CHECK(r, < 0, r);
+    SOL_EXP_CHECK_GOTO(no_more_space, send);
+
+    mdata->value_prefix_len = col;
+    if (mdata->base.state.val < 0)
+        mdata->value_prefix_len++;
+
+    if (mdata->base.state.val >= 0)
+        r = snprintf(format, sizeof(format), "%%0%" PRIu8 "" PRId32 "",
+            mdata->n_digits);
+    else
+        r = snprintf(format, sizeof(format), "%% 0%" PRIu8 "" PRId32 "",
+            mdata->n_digits + 1);
+    if (r < 0 || r > (int)sizeof(format))
+        return -EINVAL;
+
+#pragma GCC diagnostic ignored "-Wformat-nonliteral"
+    r = asprintf(&it_value, format, mdata->base.state.val);
```

Comment:
Why not simply use `%.*d` to specity the number of digits and move the `asprintf()` call to the previous `if` block? Then there's no need to use a nasty `#pragma`.